src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java

```diff
@@ -0,0 +1,218 @@
+    List<TezTask> tezTasks = Utilities.getTezTasks(plan.getRootTasks());
+    ExecutionMode executionMode = getExecutionMode(mrTasks, tezTasks);
+
+    return new GenericRecordBuilder(QUERY_EVENT_SCHEMA)
+        .set("QueryId", plan.getQueryId())
+        .set("QueryText", plan.getQueryStr())
+        .set("EventType", EventType.QUERY_SUBMITTED.name())
+        .set("Timestamp", plan.getQueryStartTime())
+        .set("User", getUser(hookContext))
+        .set("RequestUser", getRequestUser(hookContext))
+        .set("ExecutionMode", executionMode.name())
+        .set("Queue", getQueueName(executionMode, conf))
+        .set("TablesRead", getTablesFromEntitySet(plan.getInputs()))
+        .set("TablesWritten", getTablesFromEntitySet(plan.getOutputs()))
+        .set("IsMapReduce", mrTasks.size() > 0)
+        .set("IsTez", tezTasks.size() > 0)
+        .set("SessionId", hookContext.getSessionId())
+        .set("InvokerInfo", conf.getLogIdVar(hookContext.getSessionId()))
+        .set("ThreadName", hookContext.getThreadId())
+        .set("ClientIpAddress", hookContext.getIpAddress())
+        .set("ClientIpAddress", hookContext.getIpAddress())
+        .set("HookVersion", HOOK_VERSION)
+        .set("HiveAddress", getHiveInstanceAddress(hookContext))
+        .set("HiveInstanceType", getHiveInstanceType(hookContext))
+        .set("LlapApplicationId", determineLlapId(conf, executionMode))
+        .set("OperationId", hookContext.getOperationId())
+        .build();
+  }
+
+  private GenericRecord getPostHookEvent(HookContext hookContext, EventStatus status) {
+    QueryPlan plan = hookContext.getQueryPlan();
+    LOG.info("Received post-hook notification for: {}", plan.getQueryId());
+
+    GenericRecordBuilder eventBuilder =
+        new GenericRecordBuilder(QUERY_EVENT_SCHEMA)
+            .set("QueryId", plan.getQueryId())
+            .set("EventType", EventType.QUERY_COMPLETED.name())
+            .set("Timestamp", clock.millis())
+            .set("User", getUser(hookContext))
+            .set("RequestUser", getRequestUser(hookContext))
+            .set("Status", status.name())
+            .set("ErrorMessage", hookContext.getErrorMessage())
+            .set("HookVersion", HOOK_VERSION)
+            .set("OperationId", hookContext.getOperationId());
+
+    JSONObject perfObj = new JSONObject();
+    for (String key : hookContext.getPerfLogger().getEndTimes().keySet()) {
+      perfObj.put(key, hookContext.getPerfLogger().getDuration(key));
+    }
+
+    eventBuilder.set("PerfObject", perfObj.toString());
+
+    return eventBuilder.build();
+  }
+
+  private static List<String> getTablesFromEntitySet(Set<? extends Entity> entities) {
+    List<String> tableNames = new ArrayList<>();
+    for (Entity entity : entities) {
+      if (entity.getType() == TABLE || entity.getType() == PARTITION) {
+        tableNames.add(entity.getTable().getDbName() + "." + entity.getTable().getTableName());
```

Comment:
The expression `entity.getTable().getDbName() + "." + entity.getTable().getTableName()` can be replaced with `entity.getTable().getCompleteName()`.

Reply:
Didn't notice during rebase. Thanks for noticing, done.